llx/builtin_ip.go

```diff
@@ -264,82 +280,177 @@ func (i IP) Cmp(other IP) int {
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
+	return IntData(int(v.Version)), 0, nil
 }
 
 func ipSubnet(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
 	if bind.Value == nil {
 		return &RawData{Type: types.Int, Error: bind.Error}, 0, nil
 	}
-
-	v := NewIP(bind.Value.(string))
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
 	return StringData(v.Subnet()), 0, nil
 }
 
 func ipPrefix(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
 	if bind.Value == nil {
 		return &RawData{Type: types.Int, Error: bind.Error}, 0, nil
 	}
-
-	v := NewIP(bind.Value.(string))
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
 	return StringData(v.Prefix()), 0, nil
 }
 
 func ipPrefixLength(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
 	if bind.Value == nil {
 		return &RawData{Type: types.Int, Error: bind.Error}, 0, nil
 	}
-
-	v := NewIP(bind.Value.(string))
-	return IntData(v.Mask), 0, nil
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
+	return IntData(v.PrefixLength), 0, nil
 }
 
 func ipSuffix(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
 	if bind.Value == nil {
 		return &RawData{Type: types.Int, Error: bind.Error}, 0, nil
 	}
-
-	v := NewIP(bind.Value.(string))
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
 	return StringData(v.Suffix()), 0, nil
 }
 
 func ipUnspecified(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
 	if bind.Value == nil {
 		return &RawData{Type: types.Int, Error: bind.Error}, 0, nil
 	}
-
-	v := NewIP(bind.Value.(string))
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
 	return BoolData(v.IP.IsUnspecified()), 0, nil
 }
 
+func ipAddress(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
+	if bind.Value == nil {
+		return &RawData{Type: types.String, Error: bind.Error}, 0, nil
+	}
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
+	return StringData(v.Address()), 0, nil
+}
+
+func ipCIDR(e *blockExecutor, bind *RawData, chunk *Chunk, ref uint64) (*RawData, uint64, error) {
+	if bind.Value == nil {
+		return &RawData{Type: types.String, Error: bind.Error}, 0, nil
+	}
+	v, ok := bind.Value.(RawIP)
+	if !ok {
+		return nil, 0, errors.New("incorrect internal data for IP type")
+	}
+	return StringData(v.CIDR()), 0, nil
+}
+
+func any2ip(v any) (RawIP, bool) {
+	switch t := v.(type) {
+	case RawIP:
+		return t, true
+	case string:
+		return ParseIP(t), true
+	case int64:
```

Comment:
are we missing some types here? like `int` and others? 🤔

Reply:
I'm guessing no, but if so, we won't ever use the other types, I think.

Reply:
Not at this position, because the internal values can only be these 3, ie for int it can only be an int64. Much of the backend is built around this.